File: src/link/bundle/bundle.cpp
// File: bundle.cpp
// Project: bundle
// Created Date: 16/08/2022
// Author: Shun Suzuki
// -----
// Last Modified: 08/01/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2022 Shun Suzuki. All rights reserved.
//

#include "autd3/link/bundle.hpp"

#include "autd3/core/link.hpp"

namespace autd3::link {

class BundleImpl final : public core::Link {
 public:
  explicit BundleImpl(std::vector<core::LinkPtr> links) : Link(), _is_open(false), _links(std::move(links)) {}
  ~BundleImpl() override = default;
  BundleImpl(const BundleImpl& v) noexcept = delete;
  BundleImpl& operator=(const BundleImpl& obj) = delete;
  BundleImpl(BundleImpl&& obj) = delete;
  BundleImpl& operator=(BundleImpl&& obj) = delete;

  bool open(const core::Geometry& geometry) override {
    if (is_open()) return true;
    _is_open = true;
    return std::all_of(_links.begin(), _links.end(), [&geometry](const auto& link) { return link->open(geometry); });
  }

  bool close() override {
    _is_open = false;
    return std::all_of(_links.begin(), _links.end(), [](const auto& link) { return link->close(); });
  }

  bool send(const driver::TxDatagram& tx) override {
    return std::accumulate(_links.begin(), _links.end(), true, [&tx](const bool acc, const auto& link) { return acc & link->send(tx); });
  }

  bool receive(driver::RxDatagram& rx) override {
    return std::accumulate(_links.rbegin(), _links.rend(), true, [&rx](const bool acc, const auto& link) { return acc & link->receive(rx); });
  }

  bool is_open() override { return _is_open; }

 private:
  bool _is_open;
  std::vector<core::LinkPtr> _links;
};

core::LinkPtr Bundle::build() {
  core::LinkPtr link = std::make_unique<BundleImpl>(std::move(_links));
  return link;
}

}  // namespace autd3::link

File: src/autd3/src/link/bundle.rs
/*
 * File: bundle.rs
 * Project: link
 * Created Date: 14/06/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 05/09/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 *
 */

use std::{marker::PhantomData, time::Duration};

use autd3_driver::{
    cpu::{RxDatagram, TxDatagram},
    error::AUTDInternalError,
    geometry::{Device, Transducer},
    link::Link,
};

/// Link to bundle two links
pub struct Bundle<T: Transducer, L1: Link<T>, L2: Link<T>> {
    main_link: L1,
    sub_link: L2,
    _t: PhantomData<T>,
}

impl<T: Transducer, L1: Link<T>, L2: Link<T>> Bundle<T, L1, L2> {
    /// constructor
    ///
    /// Data is sent to both `main_link` and `sub_link`, but the received data is used from `main_link`. Also, the larger timeout of `main_link` and `sub_link` is used.
    ///
    /// # Arguments
    ///
    /// * `main_link` - main link
    /// * `sub_link` - sub link
    ///
    pub fn new(main_link: L1, sub_link: L2) -> Self {
        Self {
            main_link,
            sub_link,
            _t: PhantomData,
        }
    }

    pub fn link_main(&self) -> &L1 {
        &self.main_link
    }

    pub fn link_main_mut(&mut self) -> &mut L1 {
        &mut self.main_link
    }

    pub fn link_sub(&self) -> &L2 {
        &self.sub_link
    }

    pub fn link_sub_mut(&mut self) -> &mut L2 {
        &mut self.sub_link
    }
}

impl<T: Transducer, L1: Link<T>, L2: Link<T>> Link<T> for Bundle<T, L1, L2> {
    fn open(&mut self, devices: &[Device<T>]) -> Result<(), AUTDInternalError> {
        if self.is_open() {
            return Ok(());
        }

        self.main_link.open(devices)?;
        self.sub_link.open(devices)?;

        Ok(())
    }

    fn close(&mut self) -> Result<(), AUTDInternalError> {
        if !self.is_open() {
            return Ok(());
        }

        self.main_link.close()?;
        self.sub_link.close()?;

        Ok(())
    }

    fn send(&mut self, tx: &TxDatagram) -> Result<bool, AUTDInternalError> {
        if !self.is_open() {
            return Ok(false);
        }

        self.main_link.send(tx)?;
        self.sub_link.send(tx)?;

        Ok(true)
    }

    fn receive(&mut self, rx: &mut RxDatagram) -> Result<bool, AUTDInternalError> {
        if !self.is_open() {
            return Ok(false);
        }

        self.main_link.receive(rx)?;

        Ok(true)
    }

    fn is_open(&self) -> bool {
        self.main_link.is_open() && self.sub_link.is_open()
    }

    fn timeout(&self) -> Duration {
        self.main_link.timeout().max(self.sub_link.timeout())
    }
}
